refactor: Draw 와 DrawTicket 을 생성하라 (테스트코드 추가)

YHLEE9753 opened the pull request:
## 어떤 이유로 Merge 를 하셨나요?

- [ ] feature 병합
- [ ] 버그 수정
- [x] 코드 개선
- [ ] 기타(아래에 자세한 내용 기입해주세요)

## 스크린샷 및 세부 내용 - 왜 해당 Merge 가 필요한지 자세하게 설명해주세요
- 테스트 코드에서 mock 테스트 추가하였습니다.
  - mock 테스트는 코드 수정시 확인이 어려워 validation 부분의 기존 테스트 코드를 삭제하지는 않았습니다.
- controller 테스트 코드 작성하였습니다.
- 자바 표준 예외 적용을 하였습니다.
- entity not found 예외는 사용하였습니다.

close #22

YHLEE9753 committed:
feat: db 설정 추가

YHLEE9753 committed:
feat: Draw, DrawTicket validation 추가

YHLEE9753 committed:
feat: 도메인 저장에서 발생하는 에러 추가

- InvalidDrawQuantityException: draw 추첨시 재고 0 일때 발생
- TimeSequnceException: draw 저장시 등록 시간이 유효하지 않을때 발생

YHLEE9753 committed:
feat: Draw, DrawTicket 저장 로직 작성

YHLEE9753 committed:
feat: Draw, DrawTicket controller 저장 로직 작성

YHLEE9753 committed:
feat: Draw, DrawTicket 저장로직을 테스트하라

- 정상 저장 테스트진행
- validation 에서 예외가 발생하는 테스트 추가

YHLEE9753 committed:
feat: 기존 코드와 merge 진행하라
- 테스트 코드 전부 돌아가는것 확인

YHLEE9753 committed:
refactor: exception 을 ControllerAdvice 에도 적용하라

YHLEE9753 committed:
refactor: 레이어마다 dto 와 mapper 을 분리하라

YHLEE9753 committed:
refactor: controller 의 책임을 service 에 분할하라

- controller 에서 엔티티 직접 참조 못하게 설정
- service 에서 service 층의 dto 만 참조할 수 있게 설정

YHLEE9753 committed:
refactor: 테스트 코드를 수정하라

YHLEE9753 committed:
refactor: member 예외처리 메세지 수정

YHLEE9753 committed:
test: 응모, 응모권 controller restdoc 테스트하라

YHLEE9753 committed:
refactor: JPAAudit 설정을 클래스 분리하라

YHLEE9753 committed:
refactor: exception 을 자바 표준을 사용하라

- record 빌더 추가
- 공용적이지 않은 예외는 표준 사용
- EntityNotFoundException 은 일단 제거하지 않았습니다.

YHLEE9753 committed:
refactor: mock 테스트, Mvc 테스트 추가

YHLEE9753 committed:
feat: upstream 과 merge를 진행하라

goshkim commented on src/main/java/prgrms/neoike/common/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,7 @@
+package prgrms.neoike.common.config;
+
+import org.springframework.context.annotation.Configuration;
+
```
EnableJpaAuditing 다시 추가해주시면 감사하겠습니다~

goshkim commented on src/main/java/prgrms/neoike/domain/draw/Draw.java:
```diff
@@ -56,7 +55,7 @@ public Draw(
     private void validateQuantity(int quantity) {
         if (quantity < 0) {
-            throw new InvalidInputValueException("입력된 quantity 가 음수 입니다.");
+            throw new IllegalArgumentException("입력된 quantity 가 음수 입니다.");
```
한글로 작성한다면 quantity 도 한글로 수량이라고 적어주시면 더 좋을 듯 합니다.

goshkim commented on src/main/java/prgrms/neoike/domain/draw/Draw.java:
```diff
@@ -65,7 +64,7 @@ private void validateTimeOrder(LocalDateTime startDate, LocalDateTime endDate, L
 
         if (!(isStartBeforeEnd && isEndBeforeWinning)) {
-            throw new InvalidInputValueException("입력된 Date 날짜의 순서가 맞지 않습니다.");
+            throw new IllegalArgumentException("입력된 Date 날짜의 순서가 맞지 않습니다.");
```
마찬가지로 그냥 날짜 라고 적어주시고 뒤에 어떤 값이 입력됬는지도 적어주시면 나중에 예외가 발생하였을 때 처리하기 더 편할 것 같습니다~

goshkim commented on src/main/java/prgrms/neoike/service/DrawService.java:
```diff
@@ -25,8 +25,10 @@ public DrawResponse save(ServiceDrawSaveDto drawSaveRequest) {
-        return drawRepository.findById(drawId)
+    public DrawResponse findById(Long drawId) {
+        Draw draw = drawRepository.findById(drawId)
                 .orElseThrow(() -> new EntityNotFoundException("Draw 엔티티를 id 로 찾을 수 없습니다. drawId : " + drawId));
```
+ 도 좋지만 메시지 포맷을 사용하면 더 깔끔할 것 같습니다~

YHLEE9753 committed:
refactor: Controller, Mapper, Converter 네이밍 수정

- Mapper(Controller) , Converter(Service) 클래스명
- Controller 하위 dto의 클래스명은 도메인명saveRequest
- 매퍼 메서드 이름 toDto toEntity toResponse

YHLEE9753 committed:
test: 도메인 test 코드 작성 및 service mock 테스트 제거

YHLEE9753 committed:
refactor: controller의 ResponseEntity 형식 통일화

- save 201로 통일, Url도 붙이기
- request dto JsonFormat 어노테이션 추가

YHLEE9753 committed:
refactor: test 코드 BDDMockito 로 변경

YHLEE9753 committed:
refactor: test mocking 취소 및 재응모 막는 로직을 추가하라

- test 는 기존 reposiotry 사용하던 코드로 변경
- member id 로 응모권 리스트 받아오는 로직 추가
- 재응모 막는 로직 추가

YHLEE9753 committed:
refactor: test mocking 취소 및 재응모 막는 로직을 추가하라

- test 는 기존 reposiotry 사용하던 코드로 변경
- member id 로 응모권 리스트 받아오는 로직 추가
- 재응모 막는 로직 추가

YHLEE9753 committed:
style: 세미콜론 추가

YHLEE9753 committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	src/main/java/prgrms/neoike/controller/DrawController.java

YHLEE9753 committed:
style: 주석제거
